iceoryx_posh/include/iceoryx_posh/popo/sample.hpp

```diff
@@ -27,6 +28,41 @@ namespace popo
 template <typename T>
 class PublisherInterface;
 
+namespace internal
+{
+/// @brief helper struct for sample
+template <typename T>
+struct SamplePrivateData
+{
+    SamplePrivateData(cxx::unique_ptr<T>&& samplePtr, PublisherInterface<T>& publisher);
+
+    SamplePrivateData(SamplePrivateData&& rhs);
+    SamplePrivateData& operator=(SamplePrivateData&& rhs);
+
+    SamplePrivateData(const SamplePrivateData&) = delete;
+    SamplePrivateData& operator=(const SamplePrivateData&) = delete;
+
+    cxx::unique_ptr<T> samplePtr{[](T* const) {}}; // Placeholder. This is overwritten on sample construction.
```

Comment:
Here I discovered multiple issues.
1. The `cxx::unique_ptr` should get a public member alias `using DeleteCallback = cxx::function_ref<void(T*)>;` and in this line you should use the default ctor of the `cxx::unique_ptr` (after you have handled point 2 ;) ).
2. The `cxx::unique_ptr` should get a default ctor where it is constructed with an empty `DeleteCallback` and behaves exactly like the stl counterpart. This is required since the `cxx::unique_ptr` is movable so that the source of the move operation is exactly in the same state as being default constructed - does not contain any pointer and has an unset `DeleteCallback`.
3. The `cxx::unique_ptr` has to always verify before calling the `DeleteCallback` that this callback is set, otherwise we call an unset `DeleteCallback` which is undefined behavior.
4. The `cxx::unique_ptr` constructor where you can set the `DeleteCallback` should always get a setted function_ref. We could do this in a straight forward way and just test the `DeleteCallback` if it is set in the ctor - but what do we do if its not set?! Maybe the solution is something like this:
```cpp
class always_valid_function_ref : public function_ref {
   public:
       always_valid_function_ref() =delete; // delete default ctor to ensure that it is set up
       operator function_ref(); // conversion operator to function_ref so that the user doesn't have to do a thing
};
```

Reply:
I removed the dummy initialization of the pointer and initialized it directly in the initializer list of the ctor. All the other issues are not strictly related to the `Sample` and should therefore be done in a separate issue. This PR is just a refactoring of the sample to be able to implement the custom header.

Shall I create an issue for the `unique_ptr` with your points?